File: database/export_to_csv.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Script para exportar todas las tablas de MySQL a CSV con UTF-8
"""
import mysql.connector
import csv
import sys
import os

# Configuración de la base de datos
DB_CONFIG = {
    'host': 'localhost',
    'user': 'mtto_user',
    'password': 'mtto_password',
    'database': 'mtto_db',
    'charset': 'utf8mb4',
    'collation': 'utf8mb4_unicode_ci',
    'use_unicode': True
}

# Directorio de exportación
EXPORT_DIR = '/tmp/csv_exports'
os.makedirs(EXPORT_DIR, exist_ok=True)

def export_table_to_csv(cursor, table_name):
    """Exporta una tabla a CSV con UTF-8"""
    try:
        # Obtener datos de la tabla
        cursor.execute(f"SELECT * FROM {table_name}")
        rows = cursor.fetchall()
        
        # Obtener nombres de columnas
        cursor.execute(f"SHOW COLUMNS FROM {table_name}")
        columns = [column[0] for column in cursor.fetchall()]
        
        if not rows:
            print(f"⚠ Tabla {table_name} está vacía")
            return
        
        # Escribir CSV
        csv_file = os.path.join(EXPORT_DIR, f"{table_name}.csv")
        with open(csv_file, 'w', newline='', encoding='utf-8-sig') as f:
            writer = csv.writer(f, quoting=csv.QUOTE_ALL)
            # Escribir headers
            writer.writerow(columns)
            # Escribir datos
            writer.writerows(rows)
        
        print(f"✓ {table_name}: {len(rows)} registros exportados")
        return True
    except Exception as e:
        print(f"✗ Error exportando {table_name}: {e}")
        return False

def main():
    """Función principal"""
    try:
        # Conectar a MySQL
        conn = mysql.connector.connect(**DB_CONFIG)
        cursor = conn.cursor()
        
        # Obtener lista de tablas (excluyendo vistas)
        cursor.execute("""
            SELECT TABLE_NAME 
            FROM INFORMATION_SCHEMA.TABLES 
            WHERE TABLE_SCHEMA = 'mtto_db' 
            AND TABLE_TYPE = 'BASE TABLE'
            ORDER BY TABLE_NAME
        """)
        tables = [row[0] for row in cursor.fetchall()]
        
        print(f"Exportando {len(tables)} tablas a CSV...")
        print(f"Directorio: {EXPORT_DIR}\n")
        
        # Exportar cada tabla
        exported = 0
        for table in tables:
            if export_table_to_csv(cursor, table):
                exported += 1
        
        cursor.close()
        conn.close()
        
        print(f"\n✓ Exportación completada: {exported}/{len(tables)} tablas")
        print(f"Archivos guardados en: {EXPORT_DIR}")
        
    except Exception as e:
        print(f"✗ Error: {e}")
        sys.exit(1)

if __name__ == '__main__':
    main()

File: backend/scripts/exportToCSV.js
// -*- coding: utf-8 -*-
/**
 * Script para exportar todas las tablas de MySQL a CSV con UTF-8
 */
import mysql from 'mysql2/promise';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Configuración de la base de datos
const dbConfig = {
    host: process.env.MYSQL_HOST || 'localhost',
    port: parseInt(process.env.MYSQL_PORT || '3306'),
    user: process.env.MYSQL_USER || 'mtto_user',
    password: process.env.MYSQL_PASSWORD || 'mtto_password',
    database: process.env.MYSQL_DATABASE || 'mtto_db',
    charset: 'utf8mb4'
};

// Directorio de exportación - desde la raíz del proyecto
const projectRoot = path.resolve(__dirname, '../../');
const EXPORT_DIR = path.join(projectRoot, 'database/csv_exports');

// Crear directorio si no existe
if (!fs.existsSync(EXPORT_DIR)) {
    fs.mkdirSync(EXPORT_DIR, { recursive: true });
}

// Función para escapar valores CSV
function escapeCSV(value) {
    if (value === null || value === undefined) {
        return '';
    }
    const str = String(value);
    if (str.includes(',') || str.includes('"') || str.includes('\n')) {
        return `"${str.replace(/"/g, '""')}"`;
    }
    return str;
}

// Función para exportar tabla a CSV
async function exportTableToCSV(connection, tableName) {
    try {
        // Obtener nombres de columnas
        const [columns] = await connection.query(`
            SELECT COLUMN_NAME 
            FROM INFORMATION_SCHEMA.COLUMNS 
            WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
            ORDER BY ORDINAL_POSITION
        `, [dbConfig.database, tableName]);
        
        if (columns.length === 0) {
            console.log(`⚠ Tabla ${tableName} no encontrada`);
            return false;
        }
        
        const columnNames = columns.map(col => col.COLUMN_NAME);
        
        // Obtener datos
        const [rows] = await connection.query(`SELECT * FROM ${tableName}`);
        
        if (rows.length === 0) {
            console.log(`⚠ Tabla ${tableName} está vacía`);
            return false;
        }
        
        // Escribir CSV con BOM UTF-8
        const csvFile = path.join(EXPORT_DIR, `${tableName}.csv`);
        const stream = fs.createWriteStream(csvFile, { encoding: 'utf8' });
        
        // Escribir BOM UTF-8
        stream.write('\ufeff');
        
        // Escribir headers
        stream.write(columnNames.map(escapeCSV).join(',') + '\n');
        
        // Escribir datos
        for (const row of rows) {
            const values = columnNames.map(col => escapeCSV(row[col]));
            stream.write(values.join(',') + '\n');
        }
        
        stream.end();
        
        console.log(`✓ ${tableName}: ${rows.length} registros exportados`);
        return true;
    } catch (error) {
        console.error(`✗ Error exportando ${tableName}:`, error.message);
        return false;
    }
}

// Función principal
async function main() {
    let connection;
    try {
        // Conectar a MySQL
        connection = await mysql.createConnection(dbConfig);
        await connection.query('SET NAMES utf8mb4 COLLATE utf8mb4_unicode_ci');
        
        console.log('✓ Conectado a MySQL');
        console.log(`Exportando tablas a: ${EXPORT_DIR}\n`);
        
        // Obtener lista de tablas (excluyendo vistas)
        const [tables] = await connection.query(`
            SELECT TABLE_NAME 
            FROM INFORMATION_SCHEMA.TABLES 
            WHERE TABLE_SCHEMA = ? 
            AND TABLE_TYPE = 'BASE TABLE'
            ORDER BY TABLE_NAME
        `, [dbConfig.database]);
        
        const tableNames = tables.map(t => t.TABLE_NAME);
        console.log(`Encontradas ${tableNames.length} tablas para exportar\n`);
        
        // Exportar cada tabla
        let exported = 0;
        for (const tableName of tableNames) {
            if (await exportTableToCSV(connection, tableName)) {
                exported++;
            }
        }
        
        await connection.end();
        
        console.log(`\n✓ Exportación completada: ${exported}/${tableNames.length} tablas`);
        console.log(`Archivos guardados en: ${EXPORT_DIR}`);
        
    } catch (error) {
        console.error('✗ Error:', error.message);
        if (connection) {
            await connection.end();
        }
        process.exit(1);
    }
}

// Ejecutar
main();

